constellations/orion/src/controllers/classes_by_semesters_controller.go

```diff
@@ -0,0 +1,118 @@
+package controllers
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/ahsu1230/mathnavigatorSite/constellations/orion/src/domains"
+	"github.com/ahsu1230/mathnavigatorSite/constellations/orion/src/repos"
+	"github.com/gin-gonic/gin"
+)
+
+func GetAllProgramsSemestersClasses(c *gin.Context) {
+	var listResults []domains.ProgramClassesBySemester
+
+	// Fetch progrmas, semesters, classes from repo functions
+	publishedOnly := ParseParamPublishedOnly(c)
+
+	programs, err := repos.ProgramRepo.SelectAll(publishedOnly)
+	semesters, err := repos.SemesterRepo.SelectAll(publishedOnly)
+	classes, err := repos.ClassRepo.SelectAll(publishedOnly)
+
+	// Convert lists into maps
+	programMap := createProgramMap(programs)
+
+	// Loop over semesterIds
+	for i := 0; i < len(semesters); i++ {
+		// Create Semester Object
+		semesterObj := semesters[i]
+
+		programClasses := createProgramClassesForSemester(c, semesters[i].SemesterId, programs, classes, programMap)
+
+		// Make ProgramClassesBySemester struct
+		programClassesBySemester := domains.ProgramClassesBySemester{
+			Semester:       semesterObj,
+			ProgramClasses: programClasses,
+		}
+
+		listResults = append(listResults, programClassesBySemester)
+	}
+
+	if err != nil {
+		c.Error(err)
+		c.String(http.StatusNotFound, err.Error())
+	} else {
+		c.JSON(http.StatusOK, &listResults)
+	}
+}
+
+func createProgramClassesForSemester(c *gin.Context, semesterId string, programs []domains.Program, classes []domains.Class, programMap map[string]domains.Program) []domains.ProgramClass {
```

Comment:
ok so looks like this is a helper function.

You can think of this helper function as only dealing with programs, semesters, classes, etc. And will return a list of `ProgramClasses`. Which is all great.

The function that calls THIS function `GetAllProgramsSemestersClasses` will deal with the JSON stuff. So you don't need to pass the gin.Context to this function. This function will ONLY deal with the domains. The caller will deal with JSON.

That way, it's neatly separated which function does what.

Reply:
I would have this function have TWO return types (unique in Go).

It will either return a list of ProgramClasses (if everything is ok) OR return an error (if something wrong happened)